ls1intum/Artemis
# `Programming exercises`: Increase interactivity of programming feedback and grading charts

### Checklist
#### General


#### Client

### Motivation and Context


This PR increases the interactivity for the charts on programming feedback detail and programming grading pages

### Description

**Test case tab on programming grading:**
The following changes are made:

- By clicking on an arbitrary test case in the weight and bonus chart, the table next to the chart gets filtered accordingly. Additionally, a reset button appears that allows to reset the filtering
- By clicking on an arbitrary test case on the points chart, the user gets delegated to the corresponding statistics page
- The chart is now instantly translation sensitive

**Static code analysis tab on programming grading:**
The following changes are made:

- By clicking on an arbitrary category within the penalty bar, the table next to the chart gets filtered accordingly. Additionally, a reset button appears that allows to reset the filtering
- By clicking on an arbitrary category within any other bar, the user gets delegated to the corresponding scores page of the exercise
- The chart is now instantly translation sensitive

**Programming feedback details:**
The following changes have been made:

- By clicking on any bar in the chart, the feedback items below it get filtered accordingly. Additionally, a disclaimer is shown remembering the user that the feedback items are currently filtered

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise

1. Log in to Artemis as instructor
2. Navigate to the grading page of the exercise
3. Play around with the configurations. Test whether filtering of test cases and categories works as expected. Also make sure that the modifications persist if you first filter for one test case/ category, modify it and reset the filter again. Last but not least, make sure that your changes are saved appropriately after pressing save if they were done while filtering.
4. Make sure that the routing works as described above
5. Participate with the students on this exercise
6. Check the bar chart in programming feedback details. Make sure that the filtering works as expected and the disclaimer is shown if a filter is applied.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





`sca-category-distribution-chart.component.ts`: 90.32% | 88.4%
`test-case-distribution-chart.component.ts`: 93.33% | 88.6% 
`programming-exercise-configure-grading.component.ts`: 71.42% | 83.02%
`result-detail.component.ts`: 73.01% | 89.44%
### Screenshots


Example how the filtering of the feedback items can look:
![image](https://user-images.githubusercontent.com/80622272/151837984-e45ddec1-e251-4070-ab05-940675e9a34f.png)

Example how the filtering of the test cases can look:
![image](https://user-images.githubusercontent.com/80622272/151838227-887d9928-b6ae-4a09-bd86-a4a9a86c65e4.png)

Example how the filtering of the sca categories can look:
![image](https://user-images.githubusercontent.com/80622272/151838657-3131c9bc-7a98-435b-968a-58a364f1da06.png)
